File: dump_microcode.py
#!/usr/bin/env python3
"""
Intel Microcode Dump Tool
Reads and displays Intel microcode header information with checksum validation
"""

import struct
import sys
import os

class IntelMicrocodeHeader:
    def __init__(self, data: bytes):
        """Parse Intel microcode header from 48 bytes of data"""
        if len(data) < 48:
            raise ValueError(f"Insufficient data for microcode header: {len(data)} bytes, need 48")
        
        # Intel microcode header format: 12 32-bit fields (48 bytes total)
        fields = struct.unpack('<12L', data[:48])
        self.header_version = fields[0]
        self.update_revision = fields[1]
        self.date = fields[2]
        self.processor_signature = fields[3]
        self.checksum = fields[4]
        self.loader_revision = fields[5]
        self.processor_flags = fields[6]
        self.data_size = fields[7]
        self.total_size = fields[8]
        self.reserved = fields[9:12]

def bcd_to_dec(bcd: int) -> int:
    """Convert BCD (Binary Coded Decimal) to decimal"""
    return ((bcd >> 4) * 10) + (bcd & 0x0F)

def parse_date(date_value: int) -> str:
    """Parse Intel microcode date format (BCD encoded)"""
    # Extract bytes from 32-bit date value
    date_bytes = [
        (date_value >> 0) & 0xFF,   # Byte 0 → YY (lo)
        (date_value >> 8) & 0xFF,   # Byte 1 → YY (hi)
        (date_value >> 16) & 0xFF,  # Byte 2 → DD
        (date_value >> 24) & 0xFF   # Byte 3 → MM
    ]
    
    month = bcd_to_dec(date_bytes[3])      # Byte 3 → MM
    day = bcd_to_dec(date_bytes[2])        # Byte 2 → DD
    year = bcd_to_dec(date_bytes[1]) * 100 + bcd_to_dec(date_bytes[0])  # Bytes 1,0 → YYYY
    
    return f"{year:04d}-{month:02d}-{day:02d}"

def print_header(header: IntelMicrocodeHeader) -> None:
    """Print formatted microcode header information"""
    total_size = header.total_size if header.total_size else 2048
    
    print("Header Version     :", f"0x{header.header_version:08X}")
    print("Update Revision    :", f"0x{header.update_revision:08X}")
    print("Date               :", parse_date(header.date))
    print("Processor Signature:", f"0x{header.processor_signature:08X}")
    print("Checksum           :", f"0x{header.checksum:08X}")
    print("Loader Revision    :", f"0x{header.loader_revision:08X}")
    print("Processor Flags    :", f"0x{header.processor_flags:08X}")
    print("Data Size (bytes)  :", header.data_size)
    print("Total Size (bytes) :", total_size)

def validate_checksum(data: bytes) -> bool:
    """Validate Intel microcode checksum (sum of all 32-bit words should be 0)"""
    if len(data) % 4 != 0:
        # Pad to 4-byte boundary if necessary
        padding = 4 - (len(data) % 4)
        data = data + b'\x00' * padding
    
    checksum = 0
    for i in range(0, len(data), 4):
        # Unpack 32-bit little-endian value
        val = struct.unpack('<L', data[i:i+4])[0]
        checksum = (checksum + val) & 0xFFFFFFFF
    
    return checksum == 0

def main():
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <microcode_file>", file=sys.stderr)
        return 1
    
    filename = sys.argv[1]
    
    # Check if file exists
    if not os.path.exists(filename):
        print(f"Error: File '{filename}' not found", file=sys.stderr)
        return 1
    
    try:
        # Read microcode file
        with open(filename, 'rb') as fp:
            # Read header first
            header_data = fp.read(48)
            if len(header_data) < 48:
                print("Error: File too small to contain microcode header", file=sys.stderr)
                return 1
            
            # Parse header
            header = IntelMicrocodeHeader(header_data)
            total_size = header.total_size if header.total_size else 2048
            
            # Read entire microcode update
            fp.seek(0)
            full_data = fp.read(total_size)
            
            if len(full_data) < total_size:
                print(f"Error: File too small. Expected {total_size} bytes, got {len(full_data)}", file=sys.stderr)
                return 1
            
            # Print header information
            print_header(header)
            
            # Validate checksum
            is_valid = validate_checksum(full_data)
            print("Checksum Valid     :", "YES" if is_valid else "NO")
            
            return 0
            
    except IOError as e:
        print(f"Error reading file: {e}", file=sys.stderr)
        return 1
    except ValueError as e:
        print(f"Error parsing microcode: {e}", file=sys.stderr)
        return 1
    except Exception as e:
        print(f"Unexpected error: {e}", file=sys.stderr)
        return 1

if __name__ == "__main__":
    sys.exit(main())

File: dump_microcode.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>

#pragma pack(push, 1)
struct intel_microcode_header {
    uint32_t header_version;
    uint32_t update_revision;
    uint32_t date;
    uint32_t processor_signature;
    uint32_t checksum;
    uint32_t loader_revision;
    uint32_t processor_flags;
    uint32_t data_size;
    uint32_t total_size;
    uint32_t reserved[3];
};
#pragma pack(pop)

int bcd_to_dec(uint8_t bcd) {
    return ((bcd >> 4) * 10) + (bcd & 0x0F);
}

void print_header(struct intel_microcode_header *hdr) {
    uint8_t *date_bytes = (uint8_t *)&hdr->date;

    int month = bcd_to_dec(date_bytes[3]);      // Byte 0 → MM
    int day   = bcd_to_dec(date_bytes[2]);      // Byte 1 → DD
    int year  = bcd_to_dec(date_bytes[1]) * 100 // Byte 3 → YY (hi)
              + bcd_to_dec(date_bytes[0]);      // Byte 2 → YY (lo)

    printf("Header Version     : 0x%08X\n", hdr->header_version);
    printf("Update Revision    : 0x%08X\n", hdr->update_revision);
    printf("Date               : %04d-%02d-%02d\n", year, month, day);
    printf("Processor Signature: 0x%08X\n", hdr->processor_signature);
    printf("Checksum           : 0x%08X\n", hdr->checksum);
    printf("Loader Revision    : 0x%08X\n", hdr->loader_revision);
    printf("Processor Flags    : 0x%08X\n", hdr->processor_flags);
    printf("Data Size (bytes)  : %u\n", hdr->data_size);
    printf("Total Size (bytes) : %u\n", hdr->total_size ? hdr->total_size : 2048);
}

int validate_checksum(const uint8_t *buf, size_t size) {
    uint32_t sum = 0;
    for (size_t i = 0; i < size; i += 4) {
        uint32_t val = *(uint32_t *)(buf + i);
        sum += val;
    }
    return (sum == 0);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <microcode_file>\n", argv[0]);
        return EXIT_FAILURE;
    }

    FILE *fp = fopen(argv[1], "rb");
    if (!fp) {
        perror("fopen");
        return EXIT_FAILURE;
    }

    struct intel_microcode_header hdr;
    if (fread(&hdr, sizeof(hdr), 1, fp) != 1) {
        perror("fread");
        fclose(fp);
        return EXIT_FAILURE;
    }

    uint32_t total_size = hdr.total_size ? hdr.total_size : 2048;
    uint8_t *buf = malloc(total_size);
    if (!buf) {
        perror("malloc");
        fclose(fp);
        return EXIT_FAILURE;
    }

    fseek(fp, 0, SEEK_SET);
    if (fread(buf, total_size, 1, fp) != 1) {
        perror("fread full update");
        free(buf);
        fclose(fp);
        return EXIT_FAILURE;
    }

    print_header(&hdr);
    printf("Checksum Valid     : %s\n", validate_checksum(buf, total_size) ? "YES" : "NO");

    free(buf);
    fclose(fp);
    return EXIT_SUCCESS;
}

